### PR TITLE
adding cudnn V6 support

### PR DESCRIPTION
**Pull request by ngimel**
cudnn V6 rc is now available. This PR adds dilated convolution support via cudnn, and updates rnn calls for the new version. I understand it should not go on master branch, so let me know how best to handle it.

### CODE REVIEW

**Comment by apaszke** [comment]
It would be the easiest if we could just merge it, have it coexist with cuDNN5 code, and choose the mode based on the library we find. But it seems to me that it would mess up the code pretty badly in many places. 😕
If you're willing to change that, I think we can merge it into master, if not, we can probably put it in cudnn_v6 branch and rebase it periodically. @soumith? @colesbury?

**Comment by soumith** [comment]
when is v6 final release date :)
Depending on that, we can plan either ways.

**Comment by ngimel** [comment]
It _should_ be some time in February. @apaszke, I don't think having them coexist is that difficult - there's no harm in having dilation parameter in convolution routines for v5, as long as it is always 1:-). Other changes are minimal. I'll work on a modified PR.

**Comment by apaszke** [comment]
It's not difficult, it's just annoying 😉 Like having additional fields in ctypes structs, that are conditionally included. But of course it's not a big deal, if you change that I think we can merge it

File: torch/csrc/cudnn/Descriptors.h [hunk 1]
```diff
@@ -62,8 +62,8 @@ struct ConvolutionDescriptor
   ~ConvolutionDescriptor() {
     cudnnDestroyConvolutionDescriptor(desc);
   }
-  void set(cudnnDataType_t dataType, int dim, int* pad, int* stride) {
-    int upscale[3] = {1, 1, 1};
+  void set(cudnnDataType_t dataType, int dim, int* pad, int* stride, int * upscale) {
+//    int upscale[3] = {1, 1, 1};
```
Comment (soumith) [inline torch/csrc/cudnn/Descriptors.h]: minor: can you remove commented code

**Comment by ngimel** [comment]
Actually, Dockerfile-v6 will work only with this PR - can you please merge it?

**Comment by apaszke** [comment]
@ngimel We have a report on slack that the main Dockerfile fails too. Can you please take a look at that?

**Comment by ngimel** [comment]
Looks like he has solved his problems (his last message is nvm).

File: torch/csrc/cudnn/Descriptors.h [hunk 2]
```diff
@@ -62,8 +62,7 @@ struct ConvolutionDescriptor
   ~ConvolutionDescriptor() {
     cudnnDestroyConvolutionDescriptor(desc);
   }
-  void set(cudnnDataType_t dataType, int dim, int* pad, int* stride) {
-    int upscale[3] = {1, 1, 1};
+  void set(cudnnDataType_t dataType, int dim, int* pad, int* stride, int * upscale) {
```
Comment (apaszke) [inline torch/csrc/cudnn/Descriptors.h]: Why is it `upscale` here and `dilation` everywhere else?

**Review by apaszke — APPROVED** [review]

File: torch/csrc/cudnn/Descriptors.h (hunk 2 above)
Comment (ngimel) [inline torch/csrc/cudnn/Descriptors.h]: No particular reason. It was called upscale here before, and cudnn.h the corresponding parameter is called dilation, so I used dilation everywhere without thinking much about it.

**Comment by apaszke** [comment]
@ngimel can you please resolve the conflicts? I think we can merge it in afterwards.

**Comment by soumith** [comment]
@ngimel all CPU builds are failing. It looks like some part of the initialization was not guarded properly?

**Comment by ngimel** [comment]
Sorry about that, cudnn.version() call was unguarded.

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thanks!

**Pull request merged**